[ENH] MAPA forecaster

phoeenniixx opened the pull request:
#### Reference Issues/PRs

follows on #7101, fixes https://github.com/sktime/sktime/issues/7698




<!--
Thanks for contributing!
-->

phoeenniixx commented:
Hi @fkiraly, I have made some changes on the MAPA, please review it, made  changes to `_decompose`, `_aggregate` and have added the tests as well

fkiraly reviewed (changes requested):
Nice! It seems like it passes the tests now!

* the `cwt` related tests are not related to this PR and can be ignored.
* The failure related to `statsmodels` is a problem that we should address - the default depends on `statsmodels` while `sktime` does not. I would suggest to make the `NaiveForecaster` or a similar forecaster the default, to avoid a default dependency.

fkiraly commented on sktime/tests/test_mapa.py:
```diff
@@ -0,0 +1,213 @@
+
+            # Check if returned data matches input
+            assert isinstance(y_decomp, pd.DataFrame)
+            assert y_decomp.equals(y)
```
maybe a silly question - if `y_decomp` always matches `y`, why is it an output? We could just use `y`.

fkiraly commented on sktime/tests/test_mapa.py:
```diff
@@ -0,0 +1,213 @@
+    """
+    test_params = MAPAForecaster.get_test_params()
+
+    for params in test_params:
```
similar, suggest pytest `mark.parametrize`

fkiraly commented on sktime/forecasting/mapa.py:
```diff
@@ -0,0 +1,506 @@
+        sp=6,
+        weights=None,
+    ):
+        self.aggregation_levels = (
```
have a look at the extension template - init parans should be written to self and never changed.

phoeenniixx replied:
please check my recent commits, is this want you want?

fkiraly commented on sktime/forecasting/mapa.py:
```diff
@@ -0,0 +1,506 @@
+            isinstance(level, int) and level > 0 for level in aggregation_levels
+        ):
+            raise ValueError("All aggregation levels must be positive integers")
+        self.base_forecaster = (
```
if a forecaster is passed, you should clone it.

fkiraly commented on sktime/forecasting/mapa.py:
```diff
@@ -0,0 +1,506 @@
+                "weights": [0.5, 0.3, 0.2],
+            },
+        ]
+
```
we should also a parameter set that has as many defaults as possible.
In this case it would be the "all defaults" setting, `{}`.

fkiraly reviewed (changes requested):
Great! It looks like the forecaster now runs!
I also assume adding the tests helped ensure that it does.

I left some comments above, blocking:
* we should also test the default settings, they would lead to some errors at the current state
* a reference describing the algorithm should be added, can be a paper or R package reference.

Some further non-blocking comments on testing, above.

phoeenniixx commented:
> - a reference describing the algorithm should be added, can be a paper or R package reference.

Actually I have although added the reference to the R package but i might need some assistance with implementing the exact algo in the package as it is quite difficult for me to understand, although I did as much as i could, but it would be really helpful if someone can review it and suggest the changes (if required) to this implementation

phoeenniixx commented:
there is a problem with test_mapa:
how to import exp_smoothing there? should i do the `_check_soft_dependencies()` check there as well? or is there some other way to import in tests?

fkiraly commented:
> there is a problem with test_mapa: how to import exp_smoothing there? should i do the `_check_soft_dependencies()` check there as well? or is there some other way to import in tests?

I think it is because you are raising an error before `super().__init__` is called, in `MAPAForecaster.__init__` - try putting the `super()` call ahead of anything that is not writing of parameters to `self`.

phoeenniixx commented:
Hi @fkiraly, Thanks for the reply!
I tried calling `super.__init__()` before everything else, still the same error. Maybe I missed something?

fkiraly commented:
No, I think it is a different error now - you are referring to `ExponentialSmoothing` before having imported or otherwise defined it.

phoeenniixx commented:
So, what should i do? 
tried a dummy class (didn't work ig).
Need a little help here.

fkiraly commented:
> So, what should i do?

There are some general best practice recommendations:

* ensure to keep your branch updated from main by merging main regularly, see https://www.sktime.net/en/latest/developer_guide/git_workflow.html
* get familiar with how to use traceback and debugging to identify the root cause of exceptions - do not guess and fudge if stuck but drill down
* follow the extension template and read it, e.g., have the `self` writes before the `super().__init__()` call as the extension template shows you, see https://www.sktime.net/en/latest/developer_guide/add_estimators.html

fkiraly commented:
hm, can you be more specific which test fails and how? You may have to look at the CI logs for the traceback there.

fkiraly commented:
Also, I think you fixed it - you had a missing skip decorator, that seems to have been the problem!

fkiraly commented on sktime/forecasting/tests/test_mapa.py:
```diff
@@ -178,6 +178,18 @@ def test_aggregate(level, agg_method, expected):
         "decompose_type": "multiplicative",
     },
 ]
+from sktime.utils.dependencies._dependencies import _check_soft_dependencies
```
I think this last commit was not needed, since the decorator already took care of the conditional skip.

phoeenniixx replied:
ok! i was just being extra cautious this time

phoeenniixx replied:
I will remove it.

phoeenniixx commented:
> Also, I think you fixed it - you had a missing skip decorator, that seems to have been the problem!

yeah i found it when i rechecked everything!

phoeenniixx commented:
when i tried removing the _check_soft_dep check and added the exp_smoothing directly to `base_forecaster_params`, it again threw the same error as i had to import the exp_smoothing without the check, so i added back the _check_soft_dep check

fkiraly reviewed (changes requested):
Great!

Failures seem to be addressed.

Could you explain the parameters of the forecasters in the docstring, e.g., how they map onto mathematical objects in the reference, or, more generally, what they do?

phoeenniixx commented:
I have updated the docstring, added how the algo works and also made parameters more clearer

phoeenniixx commented:
> Failures seem to be addressed.

I have made sure now the `statsmmodels` import doesnot cause any problems. Can you please tell me which "failures" you are talking about

fkiraly reviewed (changes requested):
Nice!

Minor thing but blocking, the bullet points in the docstring will not format properly unless there are newlines before and after. You can check the docs CI element to see how it formats.

phoeenniixx commented:
I think for api-refrence, we should add it to exp_smoothing based forecasters?

phoeenniixx commented:
I think we should add the api-reference to exp_smoothing based forecasters? As it is fundamentally based on exp_smoothing, we have added other base_forecasters only because of dependency issues...

phoeenniixx commented:
Hi @fkiraly, are there any other changes required for this PR? Any algo changes or something?

fkiraly commented on sktime/forecasting/mapa.py:
```diff
@@ -0,0 +1,744 @@
+        pd.DataFrame
+            Forecasted values for the specified forecasting horizon.
+        """
+        fh = check_fh(fh)
```
did not notice this before: you do not need to call the check, that happens in the boilerplate.

fkiraly commented on sktime/forecasting/mapa.py:
```diff
@@ -0,0 +1,744 @@
+        self : MAPAForecaster
+            The fitted forecaster instance.
+        """
+        y = check_series(y)
```
you do not need to call this, happens in the boilerplate.

fkiraly reviewed (changes requested):
Looks good, I just noticed a few small things. You do not need to do input checks etc, this is handled by the boilerplate layer.

phoeenniixx commented:
Hi @fkiraly, have there been any changes in `ForecastingHorizon` ? `fh.to_absolute ` is throwing the error that `numpy.ndarray` has no attribute `to_absolute`

fkiraly reviewed (approved):
Great, I think this is ready for merge! Nice progress since the first time you worked on this!

I have not checked everything (of the logic) in minute detail, but results and spot checks look plausible.

phoeenniixx commented:
Thanks  to you for guiding me through the process, this PR helped me learn a lot about pycharm and tests!
Although I have tried my best to bring it as close as possible to R package but I'm still not sure if it is close enough to keep the R package link in the references...

fkiraly commented:
> Although I have tried my best to bring it as close as possible to R package but I'm still not sure if it is close enough to keep the R package link in the references...

I think it has served as an inspiration, so it is a valid (and imo necessary) reference.
